src/Psl/Regex/matching.php

```diff
@@ -0,0 +1,52 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\Regex;
+
+use Psl\Type\Exception\AssertException;
+use Psl\Type\Exception\CoercionException;
+use Psl\Type\TypeInterface;
+
+use function preg_match;
+
+/**
+ * Determine if $subject matches the given $pattern and return the matches.
+ *
+ * @template T of array
+ *
+ * @param non-empty-string $pattern        The pattern to match against.
+ * @param TypeInterface<T> $capture_groups What shape does the matching items have?
+ *
+ * @return T|null
+ *
+ * @throws Exception\RuntimeException        If an internal error accord.
+ * @throws Exception\InvalidPatternException If $pattern is invalid.
+ */
+function matching(string $subject, string $pattern, TypeInterface $capture_groups, int $offset = 0): ?array
```

Comment:
```suggestion
function first_match<T>(string $subject, string $pattern, TypeInterface<T> $capture_groups, int $offset = 0): T
```

Let's also add `every_match`:

```php
function every_match<T>(string $subject, string $pattern, TypeInterface<T> $capture_groups, int $offset = 0): dict<int, T>
```

see: https://3v4l.org/uNLsi

Reply:
That kinda conflicts with the capture_groups type though, since now it will be an array of strings().
Would it make sense for introducing 2 types for that as well?

Reply:
Are you ok with splitting those functions as well:

- `single_capture_groups()`
- `every_capture_groups()`

It sounds a bit crappy and is harder to 'guess' from an editor.
Another option is a `every: bool` flag which I also don't particularly like.

Reply:
no, first_match returns `T`, every_match should return `vec<T>` ( meaning we have to remap the result of preg_match_all )